\documentclass{article}

\RequirePackage{hyperref}

\usepackage[parfill]{parskip}
\usepackage[letterpaper, portrait, margin=1in]{geometry}
% Author
\usepackage{authblk}
% Images
\usepackage{graphicx}
\usepackage{lscape}
\usepackage{longtable}
% \usepackage{svg}

%\articlesubtype{This is the article type (optional)}
% \bibliography{paper-webserver.bib}

\usepackage{xcolor}
\newcommand\mwcomment[1]{\textcolor{red}{#1}}

\begin{document}

\title{K-mer based prokaryote pan-genome graph creation for long-term surveillance}

\author[1]{Kevin K Le\thanks{kle009@uottawa.ca}}
\author[1]{Victor PJ Gannon}
\author[2]{Chad R Laing\thanks{chad.laing@canada.ca}}
\affil[1]{National Microbiology Laboratory at Lethbridge, Public Health Agency of Canada, Twp Rd 9-1, Lethbridge, AB, T1J 3Z4, Canada}
\affil[2]{National Centre for Animal Diseases, Lethbridge Laboratory, Canadian Food Inspection Agency, Twp Rd 9-1, Lethbridge, AB, T1J 3Z4, Canada}

\renewcommand\Authands{ and }

\maketitle

\begin{abstract}

Current methods for pan-genome creation, which offer a callable mapping of all genes within a sample population, are based on the prior identification of known genes and produce one-off results which can not be updated as new samples are sequenced.
Compactation steps, which reduce the final output size and are common in pan-genome creation software, prevent updating the pangome after creation and individual whole-genome sequence samples from being reconstructed from a computed pan-genome.
Instead, an optimized graph structure could support traditional pan-genome queries even with a large output size.
Here we present a method for pan-genome creation based on an uncompressed De-Bruijn graph with edge metadata that functions similarly to common pan-genomes, supports updating the pan-genome in linear time, and returns source attribution during variant calling.


\end{abstract}


\section{Introduction}

% Paragraph on pan-genomes
Pan-genomes, the representation of all genes within a population, are important for distinguishing genetic components to observed phenotypes.
Traditional phylogenetic trees used to distinguish different subtypes by branch points may be poor genetic determinants in some species \cite{vernikos2015ten} because of the high degree of genetic exchange \cite{medini2005microbial}.
Instead, pan-genomes model a clade as based off \textit{core} conserved regions along with variable \textit{accesory} regions.

% Paragraph on graph structures
% text...

% Paragraph on need to constantly update
As new samples are sequenced however, the number of variants in accesory genes are expected to rise even if the core genes remain unchanged \cite{medini2005microbial}.
This is particularly relevant for some species such as \textit{Escherichia coli.} where the core genome only represents a small percentage of the overall genome \cite{fukiya2004extensive}.
Futhermore, for public health efforts and outbreak surveillance, pan-genomes data structures must be updatable in a reasonable timeframe.
The update requirement precludes a full re-computation of the pan-genome as such an approach would not scale to an ever growing dataset of genome samples.

% Paragraph for lead-in
% text...

% **************************************************************
% Keep this command to avoid text of first page running into the
% first page footnotes
\enlargethispage{-65.1pt}
% **************************************************************

\section{Related Work}


% Panseq
% PGAP ???
% Roary
% SplitMEM

\section{Implementation}

\section{Results}

\section{Discussion}


\section{Conclusions}


\textit{Conflict of interest}. None declared.

\newpage

\bibliographystyle{unsrt}
\bibliography{paper}

\end{document}
