\documentclass{article}

\RequirePackage{hyperref}

\usepackage[parfill]{parskip}
\usepackage[letterpaper, portrait, margin=1in]{geometry}
% Author
\usepackage{authblk}
% Images
\usepackage{graphicx}
\usepackage{lscape}
\usepackage{longtable}
% \usepackage{svg}

%\articlesubtype{This is the article type (optional)}
% \bibliography{paper-webserver.bib}

\usepackage{xcolor}
\newcommand\mwcomment[1]{\textcolor{red}{#1}}

\begin{document}

\title{K-mer based prokaryote pan-genome graph creation for long-term surveillance}

\author[1]{Kevin K Le\thanks{kle009@uottawa.ca}}
\author[1]{Victor PJ Gannon}
\author[2]{Chad R Laing\thanks{chad.laing@canada.ca}}
\affil[1]{National Microbiology Laboratory at Lethbridge, Public Health Agency of Canada, Twp Rd 9-1, Lethbridge, AB, T1J 3Z4, Canada}
\affil[2]{National Centre for Animal Diseases, Lethbridge Laboratory, Canadian Food Inspection Agency, Twp Rd 9-1, Lethbridge, AB, T1J 3Z4, Canada}

\renewcommand\Authands{ and }

\maketitle

\begin{abstract}

Current methods for pan-genome creation, which offer a callable mapping of all genes within a sample population, are based on the prior identification of known genes and produce one-off results which can not be updated as new samples are sequenced.
Compactation steps, which reduce the final output size and are common in pan-genome creation software, prevent updating the pangome after creation and individual whole-genome sequence samples from being reconstructed from a computed pan-genome.
Instead, an optimized graph structure could support traditional pan-genome queries even with a large output size.
Here we present a method for pan-genome creation based on an uncompressed De-Bruijn graph with edge metadata that functions similarly to common pan-genomes, supports updating the pan-genome in linear time, and returns source attribution during variant calling.


\end{abstract}


\section{Introduction}

% Paragraph on pan-genomes
Pan-genomes, the representation of all genes and variants within a population, are important for distinguishing genetic components to observed phenotypes.
In addition, bacterial pan-genomes have broad applications from estimating and characterizing the genetic diversity of a species \cite{medini2005microbial}, to pathogenic analysis \cite{tettelin2005genome, rasko2008pangenome} and drug targeting \cite{muzzi2007pan}.
Bacteria present a unique challenge to pan-genome applications due to the high degree of genetic exchange within some species \cite{medini2005microbial}.
Unlike a traditional phylogenetic tree structure, a pan-genome must be able to characterise a species based on a varying degree of shared genetic elements where as a phylogeny based on bifurcating trees might be based off a small amount of differing homologs as characterised by branch points \cite{vernikos2015ten}.
% Paragraph on graph structures
% text...

As a result of this high degree of intra-species genetic diversity, early studies \cite{tettelin2005genome, tettelin2008comparative} focused on estimating the number of genomes that should be sequenced to represent a given species of bacteria.
This led to the concept of open and closed genomes, where sequencing additional genomes lead to newly discovered genes in open genomes and not so in closed genomes \cite{tettelin2008comparative}.
Yet, these bacterial species continue to evolve and we are increasingly concerned with the development of antimicrobial resistance.
Since the term \textit{pan-genome} was originally coined, there has been an explosion in the number of complete genomes sequenced \cite{vernikos2015ten}.


Pan-genomes may represent a novel avenue for monitoring resistance development \cite{tettelin2005genome} and drug targeting \cite{muzzi2007pan}.
Currently, the term \textit{pan-genome} commonly refers to the collection of core regions which are shared within a species and accessory regions which differ within a species \cite{tettelin2005genome, vernikos2015ten}.
Species with a high degree of genetic variation often have a small fraction of their genome within their core region \cite{rouli2015bacterial} and it is often the accesory region where antimicrobial resistance mechanisms reside \cite{muzzi2007pan}.
Software methods for pangenome creation are inundated with the combinatorial number of comparisons that must be made as the number of samples increase.
Moreover, as new samples are sequenced the number of variants in accesory genes are expected to rise even if the core genes remain unchanged \cite{medini2005microbial}.

% Paragraph of what we're doing
Here we present ....

% Paragraph on need to constantly update
Futhermore, for public health efforts and outbreak surveillance, pan-genomes data structures must be updatable in a reasonable timeframe.
The update requirement precludes a full re-computation of the pan-genome as such an approach would not scale to an ever growing dataset of genome samples.

% **************************************************************
% Keep this command to avoid text of first page running into the
% first page footnotes
\enlargethispage{-65.1pt}
% **************************************************************

\section{Related Work}

% Alginment based
% Panseq
% ???
A core focus of pan-genomics software is on the identification of core and accessory gene regions, with improvements trying to increase possible sample size or accomadating greater variation and differing reading frames.
Early software such as Panseq \cite{laing2010pan} and PGAP \cite{zhao2011pgap} are based on alignment algorithmns, with core regions identified through sequence alignment and similarity.
Subsequent packages such as PanOCT \cite{fouts2012panoct} and GET_HOMOLOGUES \cite{contreras2013get_homologues} use alignment approaches for accepting sequence similarity along with clustering approaches for core region determination.
In all these cases the creation of a pan-genome follows a combinatorial increase in the number of comparisons with a larger sample size \cite{vernikos2015ten}.

Efforts to scale pan-genome creation software have focused on sampling strategies to try and fully characterise a species with a minimum number of samples.
For example, PanGP \cite{zhao2014pangp} proposes a sampling algorithmn based on the genetic diversity of species.
In place of a sampling strategy, Prariedog takes cues from genome assembly software and focuses on maintaining high throughput, with graph connectivity metrics used to differentiate core and accessory regions.

% Data structure
% SplitMEM

% Key problems and our approach
% Coding vs non-coding
The current spate of pan-genomics software have focused exclusively on coding regions of DNA.
While useful for reducing the size of input data for comparisons, evidence \cite{} has pointed to the contribution of non-coding regions to species phenotypes.


\section{Implementation}

\section{Results}

\section{Discussion}

Pan-genomes have been widely used to characterise the genetic content of a species.
Yet, whether bacteria exist as in the traditional definition of a \textit{species} has come under question due to the high rates of horizontal gene transfer and homologous recombination \cite{fraser2009bacterial}.
Pan-genomics could instead present different species as overlaying graphs thereby allowing the sub-selection of a pan-genome for a particular species while encoding the shared genetic elements.
This graph-subgraph approach could be further extended to metagenomic samples including ecological datum, and perhaps provide a framework for efforts \cite{caro2012bacterial} to delinate bacterial species based on natural populations.

% Paragraph on typing, MLST

% Something maybe we should try
For \textit{Campylobacter jejuni}, when samples were tagged with their MLST type as generated by ???? Prariedog identified???? additional shared sequences in the core??/accessroy?? region with median length ?? and maximum length ???. % Very well might not find anything, but worth checking out?
While MLST utilizes house-keeping genes found in core regions, accessory regions are often import for ecological specificity and resistance mechanisms \cite{fraser2009bacterial}.
%% Non-coding regions??

\section{Conclusions}


\textit{Conflict of interest}. None declared.

\newpage

\bibliographystyle{unsrt}
\bibliography{paper}

\end{document}
